### ISSUE TITLE
"Commonly Used" list in the first-time-opened Command Palette 

### ISSUE THREAD

**Issue by digitarald** [microsoft/vscode#169091]
In user research Command Center (#149445) we found that the Command Palette on first opening is overwhelming, which might keep users from finding its value.

One core problem is that it is sorted A-Z, the top-most commands mostly not making sense in an environment that might not even have a file open:

<img width="300" alt="image" src="https://user-images.githubusercontent.com/8599/207519476-d060650a-fc65-4aee-940f-c11df2e88f9e.png">

One idea would be to curate ~5 _commonly used_ commands that have high usage (format) and make make sense in a workspace-less state (split editor, toggle sidebars). It would be only shown when the user doesn't have past command history (which will then show the _recently used_).

cc @daviddossett @jrieken

**Comment by jrieken** [comment]
fyi @lramos15 @bpasero

**Comment by bpasero** [comment]
I think that makes sense, similar to the settings editor:

<img width="629" alt="image" src="https://user-images.githubusercontent.com/900690/207646745-1d32ff81-f992-4d69-b5cd-47acdee14eb1.png">

I am just not sure how to present this, should it be another group below the recently used commands? And how do we pick the commands to add there, based on usage numbers?

**Comment by TylerLeonhardt** [comment]
@sbatten and I were thinking:

* last command at the top
* list of pinned commands (that default to some list we come up with)
* rest of recently used
* other commands

It might be a bit busy, but it does:
* still keep the command you spam at the top
* but makes the pinned commands reachable from the keyboard

**Comment by daviddossett** [comment]
Here's what a relatively basic version of this could look like:
- "commonly used" section initially shows at the top for new installs
- After commands have been run, "recently used" pushes "commonly used" down
- Users can remove commonly used command individually, eventually removing the section

https://user-images.githubusercontent.com/25163139/207692539-4e96ac15-5789-4da9-8f6c-2521543b9c31.mp4

I'm interested in the idea of mixing in default items to "recently used" as a way to cut down the number of sections. However neither title really captures what the section is in that case. It would be strange to open the command palette for the first time and see "recently used" commands. It would also feel weird to have run a bunch of commands and see them under "commonly used".

**Comment by bpasero** [comment]
+1 for the idea of renaming "recently used" to "commonly used" and simply seeding that list with some defaults based on telemetry so that first time users get this. We could also think about filling the set up with the commonly used ones even for those users that only have a few commands that they had invoked.

In other words, I would not introduce a third group for this.

**Comment by sbatten** [comment]
We can't use the term quick access but it reminds me of this feature in Windows File Explorer. You get a default set of pinned folders and you can unpin and add more. 
<img width="185" alt="image" src="https://user-images.githubusercontent.com/6561887/207934910-89ab28e8-feaa-4f3b-9df6-1993152a59fb.png">

**Comment by TylerLeonhardt** [comment]
For our initial attempt at implementing this, we will rename `recently used` to `suggested` (or something else) and follow this pattern:

**For a new user**

They will see 5 pre-seeded items underneath `suggested`, these items will have a certain `weight` to them and will be 5 commands we recommend for new users.

When they run a command, that command will show at the very top, under the `suggested` section, but above the 5 pre-seeded commands... it now has a `weight`, or `frequency`, of `1`. If it's run again, that number bumps to `2`, etc.

**Eventually**

The first 5 items in the `suggested` section will be the last 5 run commands. This is to not mess with muscle memory, and allow folks an easy way to re-run the last command or last few commands they just ran.

The next 5 items will be the top 5 unique commands by `weight`/`frequency`. These may include pre-seeded ones or ones that that user just runs a lot.

The next X items will be the rest of the `recently used` commands as it works today. This is to allow searching for a command to still be sorted by recently used.

Then the `other commands` section remains as is.

@digitarald will be getting me a nice list of pre-seeded commands based on our telemetry and other things.

**Interesting details**

Since commands showing up in the command palette is context driven (you can't `git: commit` if you don't have a repo open) we should handle the case for when there isn't a resolvable top 5. This would happen if the pre-seeded commands are run recently and are showing in the first 5 recently run section... the next 5, which should be the next *unique* top 5 commands by `weight`/`frequency` might not have anything to pull from... or the next highest frequency is like... 0 or 1. In this case, we can just only as many as we have.

**Comment by TylerLeonhardt** [comment]
I have playing around with this proposal, and frankly, I don't like it. It might be helpful for new users to see the 5 recommended commands, but for experienced users, it makes no difference and it doesn't feel any different than the current solution.

So I have a different proposal.

For new users, they will see a new section called `commonly used`:
![image](https://user-images.githubusercontent.com/2644648/212394027-2f328314-298f-4c24-9601-7b042124dd01.png)
This is a list that we come up with based on telemetry.

As they start to run commands, recently used slowly pushes that list down below the fold and if a command inside of it is run, it will jump up to the recently used section:
![image](https://user-images.githubusercontent.com/2644648/212394328-9dc6e594-db44-4a05-953b-6e1af89d6f0c.png)

Eventually, an experienced user will either:
* no longer see the list
* have run all the commands in the list causing them to go up to `recently used`
![image](https://user-images.githubusercontent.com/2644648/212394458-fea9486c-da7b-429a-98f6-f8b35ff4f8aa.png)

This behavior is to help new users _become_ experienced users. But experienced users should have some benefit of this work, which is why I propose that New Windows swap the `commonly used` list and the `recently used` list:
![image](https://user-images.githubusercontent.com/2644648/212394660-f4f6fb1f-d5b9-4b31-9bc6-a5d56d446ecc.png)

As the screenshot says, after the first command is run in a window, the `commonly used` and `recently used` swap going back to the normal experience above.

This proposal, I feel, enables new users to learn the ropes of the command palette while also making the Command Palette a useful launching pad for new windows.

Thoughts?

**Comment by bpasero** [comment]
I think I partially get it and like the things I get:
* first time users: get to see commonly used commands
* existing users: will not see commonly used commands unless there is a commonly used command they have not run yet
* recently used commands are always on top of commonly used commands (for muscle memory)

The part I do not get is the swapping thing? I am worried that changing the order of "commonly used" and "recently used" will break muscle memory. Think about users that blindly accept the first command from the picker.

**Comment by TylerLeonhardt** [comment]
@bpasero yeah the swapping thing is a little tricky. My motivation for doing it is to ensure that that last command you ran is on top. Once the user runs any command, we expect that command to be on top. However, without the swapping idea, if in new windows we show commonly used at the top always, then the recently run command will be below it.

**Comment by bpasero** [comment]
The list of recently run commands is global though, so why would the new window show anything different?

**Comment by TylerLeonhardt** [comment]
The thought was that a new window (ex. one without a workspace), the command palette could be used as a launcher in the same idea that the welcome view of the explorer is:
![image](https://user-images.githubusercontent.com/2644648/212410419-9984c084-577d-421b-a70c-5870e607450d.png)
or the Remote Indicator is:
![image](https://user-images.githubusercontent.com/2644648/212410484-6737a6aa-c1e9-4fbf-82a7-d69d17c476b5.png)

**Comment by TylerLeonhardt** [comment]
The launcher idea came out of standup. I thought it was interesting... esp since a new window doesn't have anything open, and the first text you see is "Show All Commands":
![image](https://user-images.githubusercontent.com/2644648/212411619-4cbb2e2b-0b86-4bb1-9a42-566fbc4779ba.png)

**Comment by bpasero** [comment]
This sounds to me like it could confuse users quickly as to why sometimes "recently used" is there and sometimes "commonly used".  I do not think that users (even experienced ones) could make the connection to the fact that its only for new (empty?) windows.

**Comment by TylerLeonhardt** [comment]
Perhaps the Command Center could be the difference instead of the "New Window" idea.

* Command Center always shows "commonly used" on top since it's mostly mouse driven.
* Command Palette always shows "recently used" on top for consistency

**Comment by TylerLeonhardt** [comment]
To make the "commonly used" a little bit smarter, we were wanting the list to be updated over time based on the commands that the user actually used. So in the Command Center, that "commonly used" section that showed at the top would show the top 5 commands that the user uses, rather than 5 hard coded suggestions.

**Comment by TylerLeonhardt** [comment]
So I played around with the Command Freqency idea and ultimately, I don't think it provides a ton of value. My top 5 commands were like:
* Reload Window
* GitHub Copy Permalink
* Show Running Extensions
* Configure Display Language

I might be a bit bias because I work on VS Code but idk to me these aren't so useful and frankly I'm not about to open the Command Center, click "Show and Run Commands", then click on one of my top 5. That's too many clicks.

To simplify this change, and put focus on the users we are focusing on in this issue, new users, I have simplified the proposal above to everything but the "New Window/Command Center" proposal.

In other words:

We show a suggested list initially and over time that list either:
* gets buried underneath all of the "recently opened" items that the user actually uses
* gets replaced by the item going into "recently opened"

**Comment by TylerLeonhardt** [comment]
Perhaps one day, we allow users to pin commands (https://github.com/microsoft/vscode/issues/163509) and have that experience be tied to the Command Center.

**Comment by bpasero** [comment]
@TylerLeonhardt to clarify, even for existing users, we now always show "commonly used" commands section in commands quick access? And if so, is that section below or above "recently used"?

**Comment by TylerLeonhardt** [comment]
@bpasero yes correct. And it is always below recently used
